src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -58,44 +72,116 @@ public override async Task<LSP.CompletionItem[]> HandleRequestAsync(LSP.Completi
+                LSP.CompletionParams request,
+                CompletionItem item,
+                CompletionTrigger completionTrigger) where TCompletionItem : LSP.CompletionItem, new()
+            {
+                var completeDisplayText = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix;
+                var completionItem = new TCompletionItem
                 {
-                    Label = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix,
-                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : item.DisplayText,
+                    Label = completeDisplayText,
+                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : completeDisplayText,
                     SortText = item.SortText,
                     FilterText = item.FilterText,
                     Kind = GetCompletionKind(item.Tags),
-                    Data = new CompletionResolveData { TextDocument = request.TextDocument, Position = request.Position, DisplayText = item.DisplayText },
+                    Data = new CompletionResolveData
+                    {
+                        TextDocument = request.TextDocument,
+                        Position = request.Position,
+                        DisplayText = item.DisplayText,
+                        CompletionTrigger = completionTrigger,
+                    },
                     Preselect = item.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection,
+                    CommitCharacters = GetCommitCharacters(item)
                 };
+
+                return completionItem;
+            }
+
+            static string[]? GetCommitCharacters(CompletionItem item)
+            {
+                var commitCharacterRules = item.Rules.CommitCharacterRules;
+
+                // If the item doesn't have any special rules, just use the default commit characters.
+                if (commitCharacterRules.IsEmpty)
+                {
+                    return null;
+                }
+
+                using var _ = ArrayBuilder<char>.GetInstance(out var commitCharacters);
+                commitCharacters.AddRange(CompletionRules.Default.DefaultCommitCharacters);
+                foreach (var rule in commitCharacterRules)
+                {
+                    switch (rule.Kind)
+                    {
+                        case CharacterSetModificationKind.Add:
+                            foreach (var c in rule.Characters)
+                            {
+                                if (!commitCharacters.Contains(c))
+                                {
+                                    commitCharacters.Add(c);
+                                }
+                            }
+                            continue;
+
+                        case CharacterSetModificationKind.Remove:
+                            foreach (var c in rule.Characters)
```

Comment:
It seems like order here matters, if the remove comes before the add, then the characters to remove might not exist - or does this only modify the default set?

Reply:
I believe it only modifies the default set - the code here is modeled after the commit character behavior in local Roslyn: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs#L300

Reply:
Interesting - it looks like that exits out on Add if the char matches, on Remove if the char matches, or exits immediately on replace.  We should try and keep that behavior as best as possible.

We can continue on Add and remove - we shouldn't ever get into a situation where we remove something previously added (since it only acts on the default set).  But let's exit early on Replace as that could change the default set, same as how it's done there.